Fix #2712: latex has two distinct python functions to convert dimensions

jfbu opened the pull request:
and they behave differently. This includes PR #2711. As I don't know how extensions may have used the Python functions `width_to_latex_length(length_str)` and `latex_image_length(self, width_str)` from class `LaTeXTranslator`, they are still defined. The former with earlier meaning, the latter is now wrapper to `latex_dim_to_latexdim(self, length_str)` function from class `LaTeXTranslator`. Only this is used in `latex.py`.

jfbu committed:
Fix #2712: latex has two distinct python functions to convert dimensions

and they behave differently

tk0miya commented:
Ah, sorry. The duplication is my mistake. I added `width_to_latex_length` without knowing `latex_image_length` already exists (at 4789ef89).
It was added since sphinx-1.4.2 (released at May 29), so it is not used by 3rd party extensions yet.

I will merge these function on stable branch tomorrow.

jfbu commented:
@tk0miya oh, using `git blame` I had seen that `latex_image_length` was very old, but somehow I overlooked that `width_to_latex_length` was recent addition... (from the docstring I imagined it was very old and was needed in other contexts, so I couldn't use `\sphinxpxdimen` in it). But this is very good as extensions have not had time to get used to it so we can remove it.

Then perhaps I should maintain use of old name `latex_image_length`, rather than new name `latex_dim_to_latexdim` ? I can add a commit for that and also to completely remove `width_to_latex_length`.

jfbu committed:
Fix #2712: remove ``width_to_latex_length`` now unused function

Memo: it was added at Sphinx-1.4.2 (4789ef8), but now its functionality
has been merged into ``latex_dim_to_latexdim``.

jfbu commented:
In latest commit I removed completely `width_to_latex_length`. I still rename the modified `latex_image_length` to `latex_dim_to_latexdim`, and `latex_image_length` is maintained as wrapper, if it is not important to maintain old name we can suppress it, or conversely we can also perhaps rather keep same name `latex_image_length`. I changed the name because it is not limited a priori to use for images only.

jfbu commented:
latest commit also corrected wording in CHANGES for the entries relevant to this

jfbu commented:
@tk0miya if you want to merge in stable, I prepared https://github.com/jfbu/sphinx/tree/latex_allowpx3_onstable which is rebase of the commits here on current stable (45ddb63). It contains https://github.com/jfbu/sphinx/tree/latex_allowpx2_onstable which is the PR #2711 rebased on stable (45ddb63), but PR #2711 is already merged into master. When later stable will be merge in master, entries in CHANGES at 1.5 will need to be removed as they now appear already at 1.4.5.

jfbu committed:
latex use px if width or height of image has no unit

To comply with the general behaviour explained in doc/rest.rst#Images
Formerly, latex target ignored width/height given with no unit.

jfbu commented:
I felt I should follow the user manual at http://www.sphinx-doc.org/en/stable/rest.html#images 

> if the size has no unit or the unit is pixels, the given size will only be respected for output channels that support pixels

so far, latex was considered not to support pixels, hence dimension with  no unit was ignored. In an earlier commit I used `bp`, but now I realize I should have used `px`, this is what latest commit does.

About: https://github.com/jfbu/sphinx/tree/latex_allowpx3_onstable. I added the similar commit and replaced in passing `1.5` by `1.4.5` in `version-changed` tags (config.rst, rest.rst).

Sorry to have merged #2711 in master earlier, complicates things if this (via https://github.com/jfbu/sphinx/tree/latex_allowpx3_onstable)  is merged in stable.

jfbu commented on sphinx/writers/latex.py:
```diff
@@ -1378,19 +1360,25 @@ def depart_hlistcol(self, node):
+        # px unit not natively accepted by all engines, use \sphinxpxdimen
         elif unit == "px":
-            res = "%.3f\\sphinxpxdimen" % (float(amount))
+            length_str = "%s\\sphinxpxdimen" % amount
```
I don't see why the `.3f` conversion should be needed. TeX accepts `\d*\.?\d*` in front of a dimension.

tk0miya replied:
Indeed. no conversions are needed here. +1.

tk0miya commented:
Sorry for confusing you again. what I'll do from now on is integration of `width_to_latex_length` and `latex_image_length` only. I don't support `px` unit on stable. It's only simple refactoring.

Wait a moment, it won't be long.

tk0miya commented:
I finished integration of the functions on both branches.

tk0miya commented on sphinx/writers/latex.py:
```diff
@@ -1378,19 +1360,25 @@ def depart_hlistcol(self, node):
-        match = re.match('(\d*\.?\d*)\s*(\S*)', width_str)
+        return self.latex_dim_to_latexdim(width_str)
+
+    def latex_dim_to_latexdim(self, length_str):
```
This means conversion from `latex_dim` to `latexdim`?
I prefer `convert_to_latexdim` or `rstdim_to_latexdim`.

tk0miya replied:
Note:
I noticed that this method does not use `self` argument. It is not needed to be a method.
To change this to function is much better.

jfbu replied:
1. `rstdim_to_latexdim` looks good to me.  Should there be a prefix ?
2. because of `\sphinxpxdimen` the function makes sense only if used to write output into a LaTeX file using `sphinx.sty`. Can we still define it as function at top level of `latex.py` ?

tk0miya replied:
The package name `sphinx.writers.latex` already means the classes and functions are used for LaTeX writer. I think such prefix are redundant. So I renamed it to `rstdim_to_latexdim` simply.

tk0miya commented:
Well, I'm working to refactor this now. please waita moment. I'll finish it within a hour (maybe).

jfbu commented:
sorry for this. Error was I merge PR #2711 in master early before situation with the function + class method was clarified.

tk0miya commented:
I made a new PR #2716. @jfbu Could you review it please?

jfbu commented:
Closing this as this finally got handled via PR #2716.
